core/trino-main/src/main/java/io/trino/server/AutoScaleResource.java

```diff
@@ -0,0 +1,539 @@
+ * limitations under the License.
+ */
+package io.trino.server;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.Joiner;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import io.airlift.http.client.BodyGenerator;
+import io.airlift.http.client.HttpClient;
+import io.airlift.http.client.HttpClient.HttpResponseFuture;
+import io.airlift.http.client.Request;
+import io.airlift.http.client.StaticBodyGenerator;
+import io.airlift.http.client.StatusResponseHandler;
+import io.airlift.http.client.StatusResponseHandler.StatusResponse;
+import io.airlift.log.Logger;
+import io.trino.dispatcher.DispatchManager;
+import io.trino.metadata.AllNodes;
+import io.trino.metadata.DiscoveryNodeManager;
+import io.trino.metadata.InternalNode;
+import io.trino.metadata.NodeState;
+import io.trino.server.security.ResourceSecurity;
+
+import javax.annotation.Nullable;
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.GET;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.Response;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.charset.Charset;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static com.google.common.collect.Sets.difference;
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static io.airlift.concurrent.Threads.threadsNamed;
+import static io.airlift.http.client.HttpUriBuilder.uriBuilderFrom;
+import static io.airlift.http.client.Request.Builder.preparePut;
+import static io.trino.metadata.NodeState.ACTIVE;
+import static io.trino.metadata.NodeState.DECOMMISSIONED;
+import static io.trino.metadata.NodeState.DECOMMISSIONING;
+import static io.trino.metadata.NodeState.INACTIVE;
+import static io.trino.metadata.NodeState.SHUTTING_DOWN;
+import static io.trino.server.security.ResourceSecurity.AccessType.PUBLIC;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+import static javax.ws.rs.core.HttpHeaders.CONTENT_TYPE;
+import static javax.ws.rs.core.MediaType.APPLICATION_JSON;
+import static javax.ws.rs.core.MediaType.TEXT_PLAIN;
+
+// AutoScaleResource expose coordinator endpoints to facilitate the auto-scaling of cluster.
```

Comment:
Actually it looks like different pieces of information exposed by this class should live in different places. 
Maybe we can have sth like
`/v1/nodes/` - for listing nodes states and allowing scheduling nodes "refresh" (ACTIVE <-> DECOMISSIONING) changs
`/v1/queries` - for getting aggregated queries information

Not sure if we need stats exposed by `/v1/autoscale/stats` at all.

Then we would still have all the building blocks for autoscaling solution and the in-Trino code structure would be more natural I think.

Reply:
I guess I agree in general. The way it is encapsulate things within AutoScaleResource so to be not intrusive to rest of code, and free me from identifying an alternative way before proceed and verify the whole idea. Now we can afford formalize the code under proper locations. /v1/autoscale/stats had other value during iterations but later only  start time is used as the coordinator start time, so likely can be removed or simplified.

So I will try to do following (not part of this comment as it will take some time):
1. have a file to host /v1/nodes/ and refresh
2. have a file to host /v1/queries/ 
3. identify a place to expose or get coordinator start time (so to retire /v1/autoscale/stats)

Reply:
great. Thank you.

@findepi @dain please chime in here. Maybe there is a reason I do not see, why we do not want extra internal resources? Is security a concern here?